fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBSortQueryIndexSelectionTest.java

```diff
@@ -931,4 +937,90 @@ void sortUniqueNull() {
         }
     }
 
+    /**
+     * Verify that sort that cannot be done with any index can be enabled for in-memory sort.
+     */
+    public enum SortWithoutIndexMode {
+        DISALLOWED, MEMORY, FILE
+    }
+
+    @EnumSource(SortWithoutIndexMode.class)
+    @ParameterizedTest(name = "sortWithoutIndex [mode = {0}]")
+    public void sortWithoutIndex(SortWithoutIndexMode mode) throws Exception {
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context);
+
+            for (int i = 0; i < 2000; i++) {
+                TestRecords1Proto.MySimpleRecord.Builder recBuilder = TestRecords1Proto.MySimpleRecord.newBuilder();
+                recBuilder.setRecNo((2244 * i + 1649) % 2357); // Slightly larger prime.
+                recBuilder.setNumValue2(i);
+                recBuilder.setNumValue3Indexed(i % 5);
+                recordStore.saveRecord(recBuilder.build());
+            }
+            commit(context);
+        }
+
+        RecordQuery query = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_3_indexed").greaterThanOrEquals(2))
+                .setSort(field("num_value_2"), true)
+                .build();
+
+        if (mode == SortWithoutIndexMode.DISALLOWED) {
+            assertThrows(RecordCoreException.class, () -> {
+                planner.plan(query);
+            });
+            return;
+        }
+
+        ((RecordQueryPlanner)planner).setConfiguration(((RecordQueryPlanner)planner).getConfiguration().asBuilder().setAllowNonIndexSort(true).build());
+
+        // Index(MySimpleRecord$num_value_3_indexed [[2],>) ORDER BY num_value_2 DESC
+        RecordQueryPlan plan = planner.plan(query);
+        assertThat(plan, sort(allOf(hasProperty("reverse", equalTo(true))),
+                indexScan(allOf(indexName("MySimpleRecord$num_value_3_indexed"), bounds(hasTupleString("[[2],>"))))));
+        assertEquals(256365917, plan.planHash(PlanHashable.PlanHashKind.LEGACY));
+        assertEquals(172993081, plan.planHash(PlanHashable.PlanHashKind.FOR_CONTINUATION));
+        assertEquals(748321565, plan.planHash(PlanHashable.PlanHashKind.STRUCTURAL_WITHOUT_LITERALS));
+
+        ExecuteProperties.Builder executeProperties = ExecuteProperties.newBuilder().setReturnedRowLimit(5);
+        if (mode == SortWithoutIndexMode.FILE) {
+            executeProperties.setSkip(1001);    // Skip + limit is enough to overflow memory buffer. Skips into middle of section.
+        }
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context);
+            timer.reset();
+            int i = mode == SortWithoutIndexMode.MEMORY ? 2000 : 333;
```

Comment:
Is `i` the total number of records read? A more descriptive variable name would be valuable here